README
IP-Info

The  README  is  used  to  introduce the module and provide instructions on how to install the
module, any machine dependencies it may have (for example C compilers and installed libraries)
and any other information that should be provided before the module is installed.

A  README  file is required for CPAN modules since CPAN extracts the README file from a module
distribution  so  that  people  browsing the archive can use it to get an idea of the module's
uses.  It is usually a good idea to provide version information here so that people can decide
whether fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

Alternatively, to install with Module::Build, you can use the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc IP::Info

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=IP-Info

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/IP-Info

    CPAN Ratings
        http://cpanratings.perl.org/d/IP-Info

    Search CPAN
        http://search.cpan.org/dist/IP-Info/

LICENSE AND COPYRIGHT

Copyright (C) 2011 Mohammad S Anwar

This  program  is  free  software; you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

Changes
Revision history for IP::Info

0.16  2016-12-03T05:30:00+01:00
      - Added test scripts (t/meta-json.t and t/meta-yml.t).

0.15  2016-11-24T12:20:00+01:00
      - Added key 'provides' to the Makefile.PL script.

0.14  2016-11-21T12:20:00+01:00
      - Updated Copyright year information.

0.13  2015-05-05T11:15:00+01:00
      - Added LICENSE file.

0.12  2015-01-17T01:15:10+01:00
      - Removed SIGNATURE file.
      - Updated MANIFEST.SKIP file to ignore folder 'eumm/'.
      - Updated Copyright year information.

0.11  2014-11-06T23:15:20+01:00
      - Corrected the syntax of pod document of package IP::Info.
      - Added method parameter to the pod document.

0.10  2014-11-05T19:17:10+01:00
      - Some more tidied up the pod document.

0.09  2014-11-04T21:24:11+01:00
      - Tidied up the pod document.

0.08  2014-10-31T22:35:00+01:00
      - Updated Changes file according to CPAN::Changes::Spec package.
      - Added unit test changes.t file.

0.07  2014-09-28T00:48:35+01:00
      - Re-structured the code using 'Moo'.
      - Added two placeholder for result i.e. IP::Info::Response::Network and IP::Info::Response::Location.
      - Removed the dependency on Readonly and XML::Simple. It now uses JSON.
      - Added key "META_MERGE" key on condition (ExUtils::MakeMake->VERSION(6.46).
      - Added section "REPOSITORY" to the pod document.
      - Added key "META_MERGE" to Makefile.PL

0.06  2013-01-05T01:02:03+01:00
      - Patch for RT# 82365 requested by Andrei.
      - Added unit test for exception handler.

0.05  2011-09-05T10:11:12+01:00
      - Removed the format attribute as its no longer relevant.
      - Updated the signature of method schema() and logic as it nows saves the file as well.
      - Updated unit test for the constructor 01-new.t.
      - Added new unit for the method schema().

0.04  2011-09-02T11:12:13+01:00
      - Added response handler as suggested by Neil Bowers.

0.03  2011-08-09T15:34:11+01:00
      - Made Readonly v1.03 as minimum requirement.
      - Prepared the distribution using the Module::Signature v0.68.

0.02  2011-08-05T10:15:16+01:00
      - Renamed global var $VER to $API_VER.

0.01  2011-08-01T10:00:00+01:00
      - First version, released.

lib/IP/Info.pm
package IP::Info;

$IP::Info::VERSION = '0.16';
$IP::Info::AUTHOR  = 'cpan:MANWAR';

=head1 NAME

IP::Info - Interface to IP geographic and network data.

=head1 VERSION

Version 0.16

=cut

use JSON;
use Data::Dumper;

use IP::Info::Response;
use IP::Info::Response::Network;
use IP::Info::Response::Location;
use IP::Info::UserAgent;
use IP::Info::UserAgent::Exception;

use Digest::MD5 qw(md5_hex);
use Data::Validate::IP qw(is_ipv4);

use Moo;
use namespace::clean;
extends 'IP::Info::UserAgent';

has 'base_url' => (is => 'ro', default => sub { return 'http://api.quova.com/v1/ipinfo'; });

=head1 DESCRIPTION

Quova RESTful API provides the geographic location & network data for any Internet
Protocol address in the public address space. The information includes:

=over 5

=item * Postal code, city, state, region, country, and continent

=item * Area code (US and Canada only) and time zone

=item * Longitude and latitude

=item * DMA (Designated Market Area) and MSA (Metropolitan Statistical Area)

=item * Network information, including type, speed, carrier, and registering
        organization

=back

To obtain your Quova API key and the shared secret, register your application L<here|http://developer.quova.com>.

=head1 CONSTRUCTOR

The constructor requires the following parameters as listed below:

    +---------+----------+----------------------------------------+
    | Key     | Required | Description                            |
    +---------+----------+----------------------------------------+
    | api_key |   Yes    | API Key given by Quova.                |
    | secret  |   Yes    | Allocated share secret given by Quova. |
    +---------+----------+----------------------------------------+

    use strict; use warnings;
    use IP::Info;

    my $api_key = 'Your_API_Key';
    my $secret  = 'Your_shared_secret';
    my $info    = IP::Info->new({ api_key => $api_key, secret => $secret });

=head1 METHODS

=head2 ip_address($ip_address)

If  an  IP  address  is specified in the correct format, then the call returns an
object of type L<IP::Info::Response> object which can be queried further to  look
for specific information for that IP. In case it encounters any error it'll throw
an exception.

=over 2

=item * dot-decimal e.g. 4.2.2.2

=item * decimal notation e.g. 67240450

=back

    use strict; use warnings;
    use IP::Info;

    my $api_key  = 'Your_API_Key';
    my $secret   = 'Your_shared_secret';
    my $info     = IP::Info->new({ api_key => $api_key, secret => $secret });
    my $response = $info->ip_address('4.2.2.2');

    print "Carrier: ", $response->network->carrier , "\n";
    print "Country: ", $response->location->country, "\n";

=cut

sub ip_address {
    my ($self, $ip) = @_;

    die ("ERROR: Missing parameter IP Address.") unless defined $ip;
    die ("ERROR: Invalid IP Address [$ip].")     unless is_ipv4($ip);

    my $url      = sprintf("%s/%s?apikey=%s&sig=%s&format=json", $self->base_url(), $ip, $self->api_key, $self->_sig());
    my $response = $self->get($url);
    my $content  = from_json($response->{content});

    return IP::Info::Response->new({
        ip_address => $content->{ipinfo}->{ip_address},
        ip_type    => $content->{ipinfo}->{ip_type},
        network    => IP::Info::Response::Network->new($content->{ipinfo}->{Network}),
        location   => IP::Info::Response::Location->new($content->{ipinfo}->{Location})
    });
}

=head2 schema($file_name)

Saves the XML Schema Document in the given file (.xsd file).In case it encounters
any error it will throw an exception.

    use strict; use warnings;
    use IP::Info;

    my $api_key = 'Your_API_Key';
    my $secret  = 'Your_shared_secret';
    my $info    = IP::Info->new({ api_key => $api_key, secret => $secret });
    $info->schema('User_supplied_filename.xsd');

=cut

sub schema {
    my ($self, $file) = @_;

    die ("ERROR: Please supply the file name for the schema document.") unless defined $file;

    my $url      = sprintf("%s/schema?apikey=%s&sig=%s", $self->base_url(), $self->api_key, $self->_sig());
    my $response = $self->get($url);

    open(SCHEMA, ">$file") or die ("ERROR: Couldn't open file [$file] for writing: [$!]");
    print SCHEMA $response->{content};
    close(SCHEMA);
}

sub _sig {
    my ($self) = @_;

    my $time = time;
    return md5_hex($self->api_key . $self->secret . $time);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/IP-Info>

=head1 BUGS

Please  report  any  bugs or feature requests to C<bug-ip-info at rt.cpan.org> or
through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=IP-Info>.
I will be notified and then you'll automatically be notified of  progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc IP::Info

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=IP-Info>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/IP-Info>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/IP-Info>

=item * Search CPAN

L<http://search.cpan.org/dist/IP-Info/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of IP::Info


lib/IP/Info/Response.pm
package IP::Info::Response;

$IP::Info::Response::VERSION   = '0.16';
$IP::Info::Response::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

IP::Info::Response - Response handler for the module IP::Info.

=head1 VERSION

Version 0.16

=cut

use Data::Dumper;

use Moo;
use namespace::clean;

=head1 DESCRIPTION

Response handler for the module IP::Info and exposes the response data to user.

=cut

has 'ip_address' => (is => 'rw', required => 1);
has 'ip_type'    => (is => 'rw', required => 1);
has 'network'    => (is => 'rw', required => 1);
has 'location'   => (is => 'rw', required => 1);

=head1 METHODS

=head2 ip_type()

Returns the IP Type.

=head2 ip_address()

Returns the IP address.

=head2 network()

Returns the object of type L<IP::Info::Response::Network>.

=head2 location()

Returns the object of type L<IP::Info::Response::Location>.

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/IP-Info>

=head1 BUGS

Please  report  any  bugs or feature requests to C<bug-ip-info at rt.cpan.org> or
through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=IP-Info>.
I will be notified and then you'll automatically be notified of  progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc IP::Info::Response

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=IP-Info>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/IP-Info>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/IP-Info>

=item * Search CPAN

L<http://search.cpan.org/dist/IP-Info/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of IP::Info::Response


lib/IP/Info/UserAgent.pm
package IP::Info::UserAgent;

$IP::Info::UserAgent::VERSION   = '0.16';
$IP::Info::UserAgent::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

IP::Info::UserAgent - User agent for IP::Info.

=head1 VERSION

Version 0.16

=cut

use 5.006;
use Data::Dumper;

use HTTP::Tiny;
use IP::Info::UserAgent::Exception;

use Moo;
use namespace::clean;

has 'api_key' => ( is => 'ro', required => 1 );
has 'secret'  => ( is => 'ro', required => 1 );
has 'ua'      => ( is => 'rw', default => sub { HTTP::Tiny->new(agent => "IP-Info/v1"); } );

=head1 DESCRIPTION

The helper library for L<IP::Info> module.

=head1 METHODS

=head2 get(<url>)

The method get() expects one parameter i.e. URL and returns the standard response.
On error throws exception of type L<IP::Info::UserAgent::Exception>.

=cut

sub get {
    my ($self, $url) = @_;

    my $ua = $self->ua;
    my $response = $ua->request('GET', $url);

    my @caller = caller(1);
    @caller = caller(2) if $caller[3] eq '(eval)';

    unless ($response->{success}) {
	IP::Info::UserAgent::Exception->throw({
            method      => $caller[3],
            message     => "request to API failed",
            code        => $response->{status},
            reason      => $response->{reason},
            filename    => $caller[1],
            line_number => $caller[2] });
    }

    return $response;
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/IP-Info>

=head1 BUGS

Please  report  any  bugs or feature requests to C<bug-ip-info at rt.cpan.org> or
through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=IP-Info>.
I will be notified and then you'll automatically be notified of  progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc IP::Info

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=IP-Info>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/IP-Info>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/IP-Info>

=item * Search CPAN

L<http://search.cpan.org/dist/IP-Info/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of IP::Info::UserAgent


lib/IP/Info/Response/Network.pm
package IP::Info::Response::Network;

$IP::Info::Response::Network::VERSION   = '0.16';
$IP::Info::Response::Network::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

IP::Info::Response::Network - Placeholder for Network for the module L<IP::Info::Response>.

=head1 VERSION

Version 0.16

=cut

use Data::Dumper;

use Moo;
use namespace::clean;

has 'ip_routing_type' => (is => 'ro');
has 'carrier'         => (is => 'ro');
has 'line_speed'      => (is => 'ro');
has 'asn'             => (is => 'ro');
has 'organization'    => (is => 'ro');
has 'domain_tld'      => (is => 'ro');
has 'domain_sld'      => (is => 'ro');
has 'connection_type' => (is => 'ro');

sub BUILD {
    my ($self, $param) = @_;

    $self->{domain_tld} = $param->{Domain}->{tld};
    $self->{domain_sld} = $param->{Domain}->{sld};
}

=head1 METHODS

=head2 ip_routing_type()

Returns the IP Routing Type.

=head2 carrier()

Returns the IP Carrier.

=head2 line_speed()

Returns the IP Line Speed.

=head2 asn()

Returns the IP ASN.

=head2 organization()

Returns the IP Organization.

=head2 connection_type()

Returns the IP Connection Type.

=head2 domain_tld()

Returns the IP Domain TLD.

=head2 domain_sld()

Returns the IP Domain SLD.

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/IP-Info>

=head1 BUGS

Please  report  any  bugs or feature requests to C<bug-ip-info at rt.cpan.org> or
through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=IP-Info>.
I will be notified and then you'll automatically be notified of  progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc IP::Info::Response::Network

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=IP-Info>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/IP-Info>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/IP-Info>

=item * Search CPAN

L<http://search.cpan.org/dist/IP-Info/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of IP::Info::Response::Network


lib/IP/Info/Response/Location.pm
package IP::Info::Response::Location;

$IP::Info::Response::Location::VERSION   = '0.16';
$IP::Info::Response::Location::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

IP::Info::Response::Location - Placeholder for Location for the module L<IP::Info::Response>.

=head1 VERSION

Version 0.16

=cut

use Data::Dumper;

use Moo;
use namespace::clean;

has 'country_code'=> (is => 'ro');
has 'country_cf'  => (is => 'ro');
has 'country'     => (is => 'ro');
has 'city_cf'     => (is => 'ro');
has 'city'        => (is => 'ro');
has 'postal_code' => (is => 'ro');
has 'time_zone'   => (is => 'ro');
has 'area_code'   => (is => 'ro');
has 'state_code'  => (is => 'ro');
has 'state_cf'    => (is => 'ro');
has 'state'       => (is => 'ro');
has 'continent'   => (is => 'ro');
has 'longitude'   => (is => 'ro');
has 'latitude'    => (is => 'ro');
has 'region'      => (is => 'ro');
has 'msa'         => (is => 'ro');
has 'dma'         => (is => 'ro');

sub BUILD {
    my ($self, $param) = @_;

    $self->{country}      = $param->{CountryData}->{country};
    $self->{country_cf}   = $param->{CountryData}->{country_cf};
    $self->{country_code} = $param->{CountryData}->{country_code};

    $self->{state}        = $param->{StateData}->{state};
    $self->{state_cf}     = $param->{StateData}->{state_cf};
    $self->{state_code}   = $param->{StateData}->{state_code};

    $self->{city}         = $param->{CityData}->{city};
    $self->{city_cf}      = $param->{CityData}->{city_cf};
    $self->{postal_code}  = $param->{CityData}->{postal_code};
    $self->{area_code}    = $param->{CityData}->{area_code};
    $self->{time_zone}    = $param->{CityData}->{time_zone};
}

=head1 METHODS

=head2 longitute()

Returns the IP Location Longitude.

=head2 latitude()

Returns the IP Location Latitude.

=head2 city_cf()

Returns the IP Location City CF.

=head2 city()

Returns the IP Location City.

=head2 postal_code()

Returns the IP Location Postal Code.

=head2 time_zone()

Returns the IP Location Time Zone.

=head2 area_code()

Returns the IP Location Area Code.

=head2 region()

Returns the IP Location Region.

=head2 continent()

Returns the IP Location Continent.

=head2 state_code()

Returns the IP Location State Code.

=head2 state_cf()

Returns the IP Location State CF.

=head2 state()

Returns the IP Location State.

=head2 country_code()

Returns the IP Location Country Code.

=head2 country_cf()

Returns the IP Location Country CF.

=head2 country()

Returns the IP Location Country.

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/IP-Info>

=head1 BUGS

Please  report  any  bugs or feature requests to C<bug-ip-info at rt.cpan.org> or
through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=IP-Info>.
I will be notified and then you'll automatically be notified of  progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc IP::Info::Response::Location

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=IP-Info>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/IP-Info>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/IP-Info>

=item * Search CPAN

L<http://search.cpan.org/dist/IP-Info/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of IP::Info::Response::Location


Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'IP::Info',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/IP/Info.pm',
    ABSTRACT_FROM      => 'lib/IP/Info.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'JSON'               => 0,
        'Moo'                => 0,
        'namespace::clean'   => 0,
        'Throwable'          => 0,
        'HTTP::Tiny'         => 0,
        'Digest::MD5'        => '2.36',
        'Data::Validate::IP' => '0.10',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'IP-Info-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'IP::Info'                       => { file => 'lib/IP/Info.pm'                    , version => '0.16' },
            'IP::Info::Response'             => { file => 'lib/IP/Info/Response.pm'           , version => '0.16' },
            'IP::Info::Response::Network'    => { file => 'lib/IP/Info/Response/Network.pm'   , version => '0.16' },
            'IP::Info::Response::Location'   => { file => 'lib/IP/Info/Response/Location.pm'  , version => '0.16' },
            'IP::Info::UserAgent'            => { file => 'lib/IP/Info/UserAgent.pm'          , version => '0.16' },
            'IP::Info::UserAgent::Exception' => { file => 'lib/IP/Info/UserAgent/Exception.pm', version => '0.16' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/IP-Info.git',
                web  => 'https://github.com/manwar/IP-Info',
            },
        }})
     : ()
    ),
);
